File: pluginable/src/main/java/com/bob/pluginable/MainActivity.java
package com.bob.pluginable;

import android.content.Intent;
import android.os.Environment;
import android.view.View;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import com.bob.pluginproxy.PluginManager;

import java.io.File;

public class MainActivity extends AppCompatActivity {

    public final static String PLUGIN_NAME = "plugin.apk";
    public final static String PLUGIN_PACKAGE_NAME = "com.bob.pluginable";
    public final static String PLUGIN_CLAZZ_NAME = "com.bob.pluginable.plugin.TestActivity";
    PluginManager mPluginManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.tv_content).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent();
                intent.putExtra(PluginManager.PACKAGE_NAME, PLUGIN_PACKAGE_NAME);
                intent.putExtra(PluginManager.PLUGIN_CLASS_NAME, PLUGIN_CLAZZ_NAME);
                mPluginManager.startActivity(intent);
            }
        });
        Utils.verifyStoragePermissions(this);

        PluginManager.init(getApplicationContext());
        mPluginManager = PluginManager.getInstance();
//        String pluginApkPath = Environment.getExternalStorageDirectory() +
//                File.separator + "plugins" + File.separator + PLUGIN_NAME;
        String pluginApkPath = getCacheDir() + File.separator + PLUGIN_NAME;


        mPluginManager.loadApk(pluginApkPath);
    }
}

File: app/src/main/java/com/bob/md/MainActivity.kt
package com.bob.md

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import com.bob.md.anim.AnimActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (savedInstanceState != null) {
            Log.e(this.javaClass.simpleName, "onCreate -->" + savedInstanceState.getString("test"))
        } else {
            Log.e(this.javaClass.simpleName, "onCreate -->" + "savedInstanceState == null")
        }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        Log.e(this.javaClass.simpleName, "onRestoreInstanceState -->" + savedInstanceState.getString("test"))
    }

    override fun onResume() {
        super.onResume()

        tvAnimSet.setOnClickListener{
            startActivity(Intent(this@MainActivity, AnimActivity::class.java))
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString("test", "我被销毁了")
    }
}
